File: backend/app/api/v1/portfolio.py
"""Portfolio read-only API (Phase 13, Phase B).

Read/list endpoints over the Phase A paper ledger: account summary, open/closed
positions, paper orders and equity-snapshot history. There are no write
endpoints. SAFE MODE: this API exposes paper-accounting records only; nothing
here can create an order, mutate trading state, or reach a broker.
"""

from __future__ import annotations

from typing import Annotated, Literal, cast

from fastapi import APIRouter, Query

from app.api.deps import CurrentUser, DBSession
from app.data.portfolio_reader import (
    PortfolioEquityPoint,
    PortfolioOrder,
    PortfolioPosition,
    PortfolioSummary,
    list_orders,
    list_positions,
    load_equity_history,
    load_portfolio_summary,
)
from app.schemas.portfolio import (
    PortfolioEquityPointOut,
    PortfolioOrderOut,
    PortfolioPositionOut,
    PortfolioSummaryOut,
)

router = APIRouter(prefix="/portfolio", tags=["portfolio"])

POSITION_STATUS_PATTERN = "^(OPEN|CLOSED)$"
ORDER_STATUS_PATTERN = "^(PENDING|FILLED|CANCELLED|REJECTED)$"
POSITION_LIMIT: int = 100
ORDER_LIMIT: int = 50
EQUITY_LIMIT: int = 500
MAX_LIMIT: int = 1000


def _summary_out(summary: PortfolioSummary) -> PortfolioSummaryOut:
    return PortfolioSummaryOut(
        as_of=summary.as_of,
        cash=summary.cash,
        equity=summary.equity,
        open_pnl=summary.open_pnl,
        realized_pnl=summary.realized_pnl,
        peak_equity=summary.peak_equity,
        drawdown_pct=summary.drawdown_pct,
        margin_used=summary.margin_used,
        open_positions=summary.open_positions,
        open_orders=summary.open_orders,
    )


def _position_out(row: PortfolioPosition) -> PortfolioPositionOut:
    return PortfolioPositionOut(
        id=row.id,
        order_id=row.order_id,
        decision_id=row.decision_id,
        symbol=row.symbol,
        timeframe=row.timeframe,
        side=cast(Literal["LONG", "SHORT"], row.side),
        units=row.units,
        entry_price=row.entry_price,
        entry_ts=row.entry_ts,
        stop_loss=row.stop_loss,
        take_profit=row.take_profit,
        status=cast(Literal["OPEN", "CLOSED"], row.status),
        exit_price=row.exit_price,
        exit_ts=row.exit_ts,
        exit_reason=row.exit_reason,
        gross_pnl=row.gross_pnl,
        net_pnl=row.net_pnl,
    )


def _order_out(row: PortfolioOrder) -> PortfolioOrderOut:
    return PortfolioOrderOut(
        id=row.id,
        decision_id=row.decision_id,
        symbol=row.symbol,
        timeframe=row.timeframe,
        side=cast(Literal["LONG", "SHORT"], row.side),
        order_type=cast(Literal["NEXT_OPEN"], row.order_type),
        status=cast(Literal["PENDING", "FILLED", "CANCELLED", "REJECTED"], row.status),
        units=row.units,
        requested_price=row.requested_price,
        filled_price=row.filled_price,
        costs=row.costs,
        stop_loss=row.stop_loss,
        take_profit=row.take_profit,
        filled_at=row.filled_at,
        created_at=row.created_at,
    )


def _equity_out(point: PortfolioEquityPoint) -> PortfolioEquityPointOut:
    return PortfolioEquityPointOut(
        ts=point.ts,
        cash=point.cash,
        equity=point.equity,
        open_pnl=point.open_pnl,
        realized_pnl=point.realized_pnl,
        peak_equity=point.peak_equity,
        drawdown_pct=point.drawdown_pct,
        margin_used=point.margin_used,
    )


@router.get("/summary")
async def portfolio_summary(
    session: DBSession,
    current: CurrentUser,
) -> PortfolioSummaryOut:
    """Latest account snapshot + open counts (viewer+)."""
    summary = await load_portfolio_summary(session)
    return _summary_out(summary)


@router.get("/positions", response_model=list[PortfolioPositionOut])
async def portfolio_positions(
    session: DBSession,
    current: CurrentUser,
    status: Annotated[str | None, Query(pattern=POSITION_STATUS_PATTERN)] = None,
    limit: Annotated[int, Query(ge=1, le=MAX_LIMIT)] = POSITION_LIMIT,
) -> list[PortfolioPositionOut]:
    """Paper positions, newest first (viewer+)."""
    rows = await list_positions(session, status=status, limit=limit)
    return [_position_out(r) for r in rows]


@router.get("/orders", response_model=list[PortfolioOrderOut])
async def portfolio_orders(
    session: DBSession,
    current: CurrentUser,
    status: Annotated[str | None, Query(pattern=ORDER_STATUS_PATTERN)] = None,
    limit: Annotated[int, Query(ge=1, le=MAX_LIMIT)] = ORDER_LIMIT,
) -> list[PortfolioOrderOut]:
    """Paper orders, newest first (viewer+)."""
    rows = await list_orders(session, status=status, limit=limit)
    return [_order_out(r) for r in rows]


@router.get("/equity", response_model=list[PortfolioEquityPointOut])
async def portfolio_equity(
    session: DBSession,
    current: CurrentUser,
    limit: Annotated[int, Query(ge=1, le=MAX_LIMIT)] = EQUITY_LIMIT,
) -> list[PortfolioEquityPointOut]:
    """Snapshot history, oldest first for the equity chart (viewer+)."""
    points = await load_equity_history(session, limit=limit)
    return [_equity_out(p) for p in points]

File: frontend/src/lib/portfolio.ts
import { ApiError, authFetch } from "./auth";

// --- Domain models ----------------------------------------------------------
export type PortfolioSummary = {
  as_of: string | null;
  cash: number;
  equity: number;
  open_pnl: number;
  realized_pnl: number;
  peak_equity: number;
  drawdown_pct: number;
  margin_used: number;
  open_positions: number;
  open_orders: number;
};

export type PortfolioPosition = {
  id: string;
  order_id: string;
  decision_id: string | null;
  symbol: string;
  timeframe: string;
  side: "LONG" | "SHORT";
  units: number;
  entry_price: number;
  entry_ts: string;
  stop_loss: number | null;
  take_profit: number | null;
  status: "OPEN" | "CLOSED";
  exit_price: number | null;
  exit_ts: string | null;
  exit_reason: string | null;
  gross_pnl: number | null;
  net_pnl: number | null;
};

export type PortfolioOrder = {
  id: string;
  decision_id: string | null;
  symbol: string;
  timeframe: string;
  side: "LONG" | "SHORT";
  order_type: "NEXT_OPEN";
  status: "PENDING" | "FILLED" | "CANCELLED" | "REJECTED";
  units: number;
  requested_price: number | null;
  filled_price: number | null;
  costs: number;
  stop_loss: number | null;
  take_profit: number | null;
  filled_at: string | null;
  created_at: string;
};

export type PortfolioEquityPoint = {
  ts: string;
  cash: number;
  equity: number;
  open_pnl: number;
  realized_pnl: number;
  peak_equity: number;
  drawdown_pct: number;
  margin_used: number;
};

export type PortfolioQuery = {
  status?: string;
  limit?: number;
};

function str(value: unknown): string | undefined {
  if (typeof value === "string" && value.length > 0) return value;
  return undefined;
}

function strOrNull(value: unknown): string | null {
  return str(value) ?? null;
}

function numOr(value: unknown, fallback: number): number {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  if (typeof value === "string" && value.trim() !== "") {
    const parsed = Number(value);
    if (Number.isFinite(parsed)) return parsed;
  }
  return fallback;
}

function numOrNull(value: unknown): number | null {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  if (typeof value === "string" && value.trim() !== "") {
    const parsed = Number(value);
    if (Number.isFinite(parsed)) return parsed;
  }
  return null;
}

/** Parse a date-time string; returns empty string when malformed/absent. */
function isoOr(value: unknown): string {
  const raw = str(value);
  if (raw === undefined) return "";
  const date = new Date(raw);
  if (Number.isNaN(date.getTime())) return "";
  return date.toISOString();
}

function isoOrNull(value: unknown): string | null {
  const parsed = isoOr(value);
  return parsed === "" ? null : parsed;
}

function sideOf(value: unknown): "LONG" | "SHORT" {
  return str(value)?.toUpperCase() === "SHORT" ? "SHORT" : "LONG";
}

// --- Normalizers ------------------------------------------------------------
export function normalizeSummary(raw: Record<string, unknown> | null | undefined): PortfolioSummary {
  const value = raw ?? {};
  return {
    as_of: isoOrNull(value["as_of"]),
    cash: numOr(value["cash"], 0),
    equity: numOr(value["equity"], 0),
    open_pnl: numOr(value["open_pnl"], 0),
    realized_pnl: numOr(value["realized_pnl"], 0),
    peak_equity: numOr(value["peak_equity"], 0),
    drawdown_pct: numOr(value["drawdown_pct"], 0),
    margin_used: numOr(value["margin_used"], 0),
    open_positions: numOr(value["open_positions"], 0),
    open_orders: numOr(value["open_orders"], 0),
  };
}

export function normalizePosition(raw: Record<string, unknown>): PortfolioPosition {
  const statusRaw = str(raw["status"])?.toUpperCase();
  return {
    id: str(raw["id"]) ?? "",
    order_id: str(raw["order_id"]) ?? "",
    decision_id: strOrNull(raw["decision_id"]),
    symbol: str(raw["symbol"]) ?? "",
    timeframe: str(raw["timeframe"]) ?? "",
    side: sideOf(raw["side"]),
    units: numOr(raw["units"], 0),
    entry_price: numOr(raw["entry_price"], 0),
    entry_ts: isoOr(raw["entry_ts"]),
    stop_loss: numOrNull(raw["stop_loss"]),
    take_profit: numOrNull(raw["take_profit"]),
    status: statusRaw === "CLOSED" ? "CLOSED" : "OPEN",
    exit_price: numOrNull(raw["exit_price"]),
    exit_ts: isoOrNull(raw["exit_ts"]),
    exit_reason: strOrNull(raw["exit_reason"]),
    gross_pnl: numOrNull(raw["gross_pnl"]),
    net_pnl: numOrNull(raw["net_pnl"]),
  };
}

export function normalizeOrder(raw: Record<string, unknown>): PortfolioOrder {
  const statusRaw = str(raw["status"])?.toUpperCase();
  return {
    id: str(raw["id"]) ?? "",
    decision_id: strOrNull(raw["decision_id"]),
    symbol: str(raw["symbol"]) ?? "",
    timeframe: str(raw["timeframe"]) ?? "",
    side: sideOf(raw["side"]),
    order_type: "NEXT_OPEN",
    status:
      statusRaw === "FILLED" ||
      statusRaw === "CANCELLED" ||
      statusRaw === "REJECTED"
        ? statusRaw
        : "PENDING",
    units: numOr(raw["units"], 0),
    requested_price: numOrNull(raw["requested_price"]),
    filled_price: numOrNull(raw["filled_price"]),
    costs: numOr(raw["costs"], 0),
    stop_loss: numOrNull(raw["stop_loss"]),
    take_profit: numOrNull(raw["take_profit"]),
    filled_at: isoOrNull(raw["filled_at"]),
    created_at: isoOr(raw["created_at"]),
  };
}

export function normalizeEquityPoint(raw: Record<string, unknown>): PortfolioEquityPoint {
  return {
    ts: isoOr(raw["ts"]),
    cash: numOr(raw["cash"], 0),
    equity: numOr(raw["equity"], 0),
    open_pnl: numOr(raw["open_pnl"], 0),
    realized_pnl: numOr(raw["realized_pnl"], 0),
    peak_equity: numOr(raw["peak_equity"], 0),
    drawdown_pct: numOr(raw["drawdown_pct"], 0),
    margin_used: numOr(raw["margin_used"], 0),
  };
}

// --- REST fetchers ----------------------------------------------------------
function buildQuery(query: PortfolioQuery): string {
  const params = new URLSearchParams();
  if (query.status) params.set("status", query.status);
  if (query.limit) params.set("limit", String(query.limit));
  const qs = params.toString();
  return qs ? `?${qs}` : "";
}

export async function fetchPortfolioSummary(): Promise<PortfolioSummary> {
  const row = await authFetch<Record<string, unknown>>("/api/v1/portfolio/summary", {
    method: "GET",
  });
  return normalizeSummary(row ?? {});
}

export async function fetchPortfolioPositions(
  query: PortfolioQuery = {},
): Promise<PortfolioPosition[]> {
  const rows = await authFetch<Array<Record<string, unknown>>>(
    `/api/v1/portfolio/positions${buildQuery(query)}`,
    { method: "GET" },
  );
  return (rows ?? []).map(normalizePosition);
}

export async function fetchPortfolioOrders(
  query: PortfolioQuery = {},
): Promise<PortfolioOrder[]> {
  const rows = await authFetch<Array<Record<string, unknown>>>(
    `/api/v1/portfolio/orders${buildQuery(query)}`,
    { method: "GET" },
  );
  return (rows ?? []).map(normalizeOrder);
}

export async function fetchPortfolioEquity(
  query: PortfolioQuery = {},
): Promise<PortfolioEquityPoint[]> {
  const rows = await authFetch<Array<Record<string, unknown>>>(
    `/api/v1/portfolio/equity${buildQuery(query)}`,
    { method: "GET" },
  );
  return (rows ?? []).map(normalizeEquityPoint);
}

export { ApiError };